lib/cipherstash/index/bloom_filter.rb

```diff
@@ -1,28 +1,115 @@
-require "openssl"
+require_relative "../client/error"
 
 module CipherStash
   class Index
+    # A bloom filter implementation designed to be used with the *FilterMatch index classes.
+    #
+    # @private
     class BloomFilter
+      K_MIN = 3
+      K_MAX = 16
+      K_DEFAULT = 3
+      M_DEFAULT = 256
+      VALID_M_VALUES = [32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536]
```

Comment:
I liked the idea of being able to glance at the constant for valid values. It's also convenient to use as a constant [in the specs](https://github.com/cipherstash/ruby-client/blob/2af4041633d1aa0e683d8bed08c64a6bf0bf6895/spec/cipherstash/index/bloom_filter_spec.rb#L156) without having to define another method for generating the valid values.

Do you have a preference between the two?

Reply:
My main concern is actually that this is *harder* to read than just using maths, which I guess is the opposite of your rationale.  I feel like it would be relatively easy for there to be a missing value, or a typo'd value, and it would be harder to spot that mistake than a problem in the formula to discern whether a number was valid or not (especially if it were wrapped in a method named `power_of_2?`, for instance).

Using the same constant in the tests as in the production code is actually something of an anti-pattern, too, IMO.  Let's say there *was* a typo in the list of valid values; by using the same constant in the implementation as the tests, you couldn't possibly catch that typo, because you're testing that the (erroneous) value equals the (erroneous) value.  You can, of course, take that too far the other way, and go redefining every value in your tests Just Because, but I think this particular case is a fairly clear-cut situation of "testing the implementation, not the behaviour".

I'd be more comfortable if you moved the constant into the tests, and used a different means of determining what constitutes a valid `m` value in the implementation.